zipnumcluster-ccpyspark.py

```diff
@@ -0,0 +1,263 @@
+        parser.add_argument("--output_base_url", required=False,
+                            default='my_cdx_bucket',
+                            help="destination for output")
+        parser.add_argument("--partition_boundries_file", required=False,
+                            help="Full path to a json file containing partition boundaries. if specified, and does not exist, will be created, otherwise, will be used.")
+        parser.add_argument("--num_lines", type=int, required=False,
+                            default=3000,
+                            help="number of lines to compress in each chunk")
+        parser.add_argument("--num_output_partitions", type=int, required=False,
+                            default=300,
+                            help="number of partitions/shards")
+
+    def run_job(self, session):
+        os.makedirs(self.args.output_base_url, exist_ok=True)
+        input = self.args.input_base_url + self.args.input
+        num_partitions = self.args.num_output_partitions
+        boundries_file_uri = self.args.partition_boundries_file
+        num_lines = self.args.num_lines
+        output_base_url = self.args.output_base_url
+        rdd = session.sparkContext.textFile(input).map(parse_line).filter(lambda x: x is not None)
+
+        # Cache the RDD with MEMORY_AND_DISK storage level
+        #rdd = rdd.persist(StorageLevel.MEMORY_AND_DISK)
+        #rdd = rdd.cache()
+
+        boundaries = None
+        ##logging.info(f"Boundaries file: {boundries_file_uri}")
+        if boundries_file_uri and self.check_for_output_file(boundries_file_uri):
+            ##logging.info(f"Boundaries file found, using it: {boundries_file_uri}")
+            with self.fetch_file(boundries_file_uri) as f:
+                boundaries = json.load(f)
+        else:
+            # this percent needs to be pretty small, since this collect brings data back to driver...
+            # 1/2 percent should be fine
+            samples = rdd.keys().sample(False, 0.005).collect()
+            samples.sort()
+            
+            # Ensure more even distribution by using quantiles
+            total_samples = len(samples)
+            boundaries = []
+            for i in range(1, num_partitions):
+                idx = (i * total_samples) // num_partitions
+                if idx < len(samples):
+                    boundaries.append(samples[idx])
+            
+            temp_file_name = 'temp_range_boundaries.json'
+            with open(temp_file_name, 'w') as f:
+                json.dump(boundaries, f)
+            
+            with open(temp_file_name, 'rb') as f:
+                self.write_output_file(boundries_file_uri, f)
+
+            os.unlink(temp_file_name)
+        
+        rdd = rdd.repartitionAndSortWithinPartitions(
+            numPartitions=num_partitions,
+            partitionFunc=lambda k: get_partition_id(k,boundaries),
+            keyfunc=lambda x: x[0]) \
```

Comment:
This will sort on the first character of the SURT key only. Should use the identity function (just leave the param `keyfunc` away).

Reply:
humn - OK, I had put that in there because I thought x was a tuple at that point, and [0] was the key - but based on what you're saying, does keyfunc already get passed the first element of the tuple? I will double check the docs, but, it sounds reasonable.

Reply:
yup, right you are:

this is from spark source:
```
key=lambda k_v: keyfunc(k_v[0])
```

Reply:
That's a subtle difference between [sortByKey](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.sortByKey.html) and [sortBy](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.sortBy.html). And of course, from the documentation it's even less clear how [repartitionAndSortWithinPartitions()](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.repartitionAndSortWithinPartitions.html) does behave. Ok, in Scala or Java this would just result in an Exception "Expected list or tuple, got string". With Python this is tricky.